File: wagerproof-ios-native/Wagerproof/Features/Chat/WagerBotThinkingIndicator.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels

/// Three-dot thinking indicator + cycling verb. Direct port of Honeydew's
/// `ChatV3ThinkingIndicator` adapted for WagerProof — instead of Honeydew's
/// cooking verbs we cycle a set of betting-themed phrases so the indicator
/// reads as on-brand instead of borrowed chrome.
///
/// Two pieces:
///   1. Three pulsing dots staggered by 0.15s.
///   2. Cycling italic verb text in the brand green — phrases rotate every
///      3s with a shimmer sweep so the surface reads alive.
struct WagerBotThinkingIndicator: View {
    let ui: WagerBotUiTokens

    @State private var verbIndex: Int = 0

    /// Betting-themed verbs. Cycles every 3s while the assistant is
    /// thinking. Kept short enough to fit on a single line in every
    /// dynamic type setting.
    private let verbs: [String] = [
        "crunching the numbers",
        "running the models",
        "checking the lines",
        "comparing odds",
        "weighing the matchup",
        "scanning Polymarket",
        "reading the splits",
        "looking up trends",
        "pulling predictions",
        "spotting the edge",
        "tracking the steam",
        "syncing live data"
    ]

    private func initialIndex() -> Int {
        let ms = Int(Date().timeIntervalSince1970 * 1000) % 1000
        return min(max(0, (verbs.count * ms) / 1000), verbs.count - 1)
    }

    var body: some View {
        HStack(spacing: 10) {
            WagerBotThinkingDots(color: ui.accent)
                .frame(width: 36, height: 18)
            WagerBotShimmerVerb(text: verbs[verbIndex], color: ui.accent)
                .id(verbIndex)
                .transition(.asymmetric(
                    insertion: .opacity.combined(with: .offset(y: 8)),
                    removal: .opacity
                ))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.top, 4)
        .onAppear {
            verbIndex = initialIndex()
            // Use `Timer` not `withAnimation+repeatForever`: the verb
            // swap is an opacity/offset transition, not a continuous
            // animation, so we drive the index from a timer and let
            // SwiftUI's `.transition` animate the swap.
            Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { _ in
                Task { @MainActor in
                    withAnimation(.easeInOut(duration: 0.4)) {
                        verbIndex = (verbIndex + 1) % verbs.count
                    }
                }
            }
        }
    }
}

/// Three pulsing dots — staggered phase animator. Mirrors the dot
/// behavior in `ChatV3ToolUseChip`'s `.running` glyph but rendered as a
/// row of three so the bubble's "still thinking" state reads at a glance
/// without a Lottie sprite (Honeydew uses Lottie; we don't ship one
/// here, and three dots stay snappier).
private struct WagerBotThinkingDots: View {
    let color: Color

    var body: some View {
        HStack(spacing: 5) {
            Circle().fill(color)
                .phaseAnimator([0.3, 1.0, 0.3]) { dot, phase in
                    dot.opacity(phase)
                } animation: { _ in .easeInOut(duration: 0.9) }
            Circle().fill(color)
                .phaseAnimator([1.0, 0.3, 1.0]) { dot, phase in
                    dot.opacity(phase)
                } animation: { _ in .easeInOut(duration: 0.9).delay(0.15) }
            Circle().fill(color)
                .phaseAnimator([0.3, 1.0, 0.3]) { dot, phase in
                    dot.opacity(phase)
                } animation: { _ in .easeInOut(duration: 0.9).delay(0.3) }
        }
        .frame(width: 36, height: 8)
    }
}

/// Verb text with the app's standard iOS shimmer applied to the glyphs: a
/// readable dim base with a bright copy on top, swept by the shared
/// `.shimmering()` modifier (the same shimmer vocabulary used by every
/// skeleton in the app) so the thinking line reads consistently.
private struct WagerBotShimmerVerb: View {
    let text: String
    let color: Color

    private var font: Font { .system(size: 14, weight: .medium).italic() }

    var body: some View {
        Text(text)
            .font(font)
            .foregroundStyle(color.opacity(0.45))
            .overlay {
                // Bright copy revealed in the travelling shimmer band, masked
                // to the glyphs so only the text lights up.
                Text(text)
                    .font(font)
                    .foregroundStyle(color)
                    .shimmering()
                    .allowsHitTesting(false)
            }
    }
}

#Preview {
    WagerBotThinkingIndicator(ui: WagerBotUiTokens.resolve(.dark))
        .padding()
        .background(Color.appSurface)
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/chat/WagerBotThinkingIndicator.kt
package com.wagerproof.app.features.chat

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.core.design.components.shimmering
import kotlinx.coroutines.delay

/**
 * Port of iOS `WagerBotThinkingIndicator.swift`.
 *
 * Three pulsing accent dots (staggered 0 / 0.15 / 0.3s) + a cycling italic
 * betting verb (rotates every 3s with a shimmer sweep over the glyphs).
 */
@Composable
fun WagerBotThinkingIndicator(ui: WagerBotUiTokens, modifier: Modifier = Modifier) {
    val verbs = remember {
        listOf(
            "crunching the numbers",
            "running the models",
            "checking the lines",
            "comparing odds",
            "weighing the matchup",
            "scanning Polymarket",
            "reading the splits",
            "looking up trends",
            "pulling predictions",
            "spotting the edge",
            "tracking the steam",
            "syncing live data",
        )
    }

    // Random initial index (mirrors iOS's ms-based seed).
    var verbIndex by remember { mutableIntStateOf((System.currentTimeMillis() % verbs.size).toInt()) }

    LaunchedEffect(Unit) {
        while (true) {
            delay(3000)
            verbIndex = (verbIndex + 1) % verbs.size
        }
    }

    Row(
        modifier = modifier.padding(top = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        ThinkingDots(color = ui.accent)
        AnimatedContent(
            targetState = verbIndex,
            transitionSpec = {
                (fadeIn(tween(400)) + slideInVerticallySmall()) togetherWith fadeOut(tween(400))
            },
            label = "verb",
        ) { idx ->
            ShimmerVerb(text = verbs[idx], color = ui.accent)
        }
    }
}

private fun slideInVerticallySmall() =
    androidx.compose.animation.slideInVertically(tween(400)) { it / 6 }

@Composable
private fun ThinkingDots(color: Color) {
    val transition = rememberInfiniteTransition(label = "dots")
    Row(horizontalArrangement = Arrangement.spacedBy(5.dp), verticalAlignment = Alignment.CenterVertically) {
        // Staggered opacity pulse — 0.9s each, delayed 0 / 150 / 300ms.
        listOf(0, 150, 300).forEach { delayMs ->
            val a by transition.animateFloat(
                initialValue = 0.3f,
                targetValue = 1.0f,
                animationSpec = infiniteRepeatable(
                    tween(900, delayMillis = delayMs, easing = androidx.compose.animation.core.FastOutSlowInEasing),
                    RepeatMode.Reverse,
                ),
                label = "dot$delayMs",
            )
            Box(
                Modifier
                    .size(8.dp)
                    .clip(CircleShape)
                    .alpha(a)
                    .background(color),
            )
        }
    }
}

/**
 * Verb text with a dim base + a bright shimmering copy overlaid, masked to
 * the glyphs (the shared `.shimmering()` sweep).
 */
@Composable
private fun ShimmerVerb(text: String, color: Color) {
    Box {
        Text(
            text = text,
            fontSize = 14.sp,
            fontWeight = FontWeight.Medium,
            fontStyle = FontStyle.Italic,
            color = color.copy(alpha = 0.45f),
        )
        Text(
            text = text,
            fontSize = 14.sp,
            fontWeight = FontWeight.Medium,
            fontStyle = FontStyle.Italic,
            color = color,
            modifier = Modifier.graphicsLayer { }.shimmering(),
        )
    }
}
